### PR TITLE
docs: Rolling restart topic for K8s

### PR DESCRIPTION
**Pull request by JakeSCahill**
Review deadline: 13 December

Fixes redpanda-data/documentation-private#2136

Added a separate topic for performing a rolling restart in Kubernetes.

### CODE REVIEW

File: modules/manage/pages/cluster-maintenance/rolling-restart.adoc [hunk 1]
```diff
@@ -1,6 +1,10 @@
-= Rolling Restart
-:description: How to perform a rolling restart of your Redpanda cluster.
+= Perform a Rolling Restart
+:description: Learn how to perform a rolling restart of your Redpanda cluster.
 :rolling-restart:
+:page-context-links: [{"name": "Linux", "to": "manage:cluster-maintenance/rolling-restart.adoc" },{"name": "Kubernetes", "to": "manage:kubernetes/k-rolling-restart.adoc" } ]
+
+A rolling restart involves restarting one broker at a time, while the remaining brokers in your cluster continue running. This is to minimize downtime during a full cluster restart.
+You should perform a rolling restart during operations such as configuration updates that require a restart, version upgrades or cluster maintenance.
```
Comment (greg-szrama) [inline modules/manage/pages/cluster-maintenance/rolling-restart.adoc]:
```suggestion
You should perform a rolling restart during operations such as configuration updates that require a restart, version upgrades, or cluster maintenance.
```

File: modules/manage/pages/cluster-maintenance/rolling-restart.adoc [hunk 2]
```diff
@@ -1,6 +1,10 @@
-= Rolling Restart
-:description: How to perform a rolling restart of your Redpanda cluster.
+= Perform a Rolling Restart
+:description: Learn how to perform a rolling restart of your Redpanda cluster.
 :rolling-restart:
+:page-context-links: [{"name": "Linux", "to": "manage:cluster-maintenance/rolling-restart.adoc" },{"name": "Kubernetes", "to": "manage:kubernetes/k-rolling-restart.adoc" } ]
+
+A rolling restart involves restarting one broker at a time, while the remaining brokers in your cluster continue running. This is to minimize downtime during a full cluster restart.
```
Comment (greg-szrama) [inline modules/manage/pages/cluster-maintenance/rolling-restart.adoc]:
```suggestion
A rolling restart involves restarting one broker at a time while the remaining brokers in your cluster continue running. This is to minimize downtime during a full cluster restart.
```

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 3]
```diff
@@ -0,0 +1,107 @@
+= Perform a Rolling Restart of Redpanda in Kubernetes
+:description: Learn how to perform a rolling restart of your Redpanda cluster when it's running in Kubernetes.
+:rolling-restart:
+:page-context-links: [{"name": "Linux", "to": "manage:cluster-maintenance/rolling-restart.adoc" },{"name": "Kubernetes", "to": "manage:kubernetes/k-rolling-restart.adoc" } ]
+
+A rolling restart involves restarting one broker at a time, while the remaining brokers in your cluster continue running. This is to minimize downtime during a full cluster restart.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]:
```suggestion
A rolling restart involves restarting one broker at a time while the remaining brokers in your cluster continue running. This is to minimize downtime during a full cluster restart.
```

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+Perform a rolling restart during operations such as configuration updates that require a restart, version upgrades or cluster maintenance.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]:
```suggestion
Perform a rolling restart during operations such as configuration updates that require a restart, version upgrades, or cluster maintenance.
```

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+When you run Redpanda in Kubernetes with the Redpanda Operator or the Redpanda Helm chart, your Redpanda cluster is managed as a StatefulSet where each broker runs inside its own Pod. As a result, you can perform a rolling restart using a single command with the Kubernetes API.
+
+== Prerequisites
+
+You must have the following:
+
+* xref:deploy:deployment-option/self-hosted/kubernetes/index.adoc[A Redpanda cluster running in Kubernetes].
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc:13]: Just confirming -- this is the only prerequisite we have for this operation?

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+include::upgrade:partial$rolling-upgrades/restart-impact.adoc[leveloffset=+1]
+
+== Perform a rolling restart
+
+A rolling restart involves running the following procedure on each broker, one at a time. The Redpanda Operator and the Redpanda Helm chart automate this procedure.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]: Can we link to resources on the Redpanda Operator and Redpanda Helm chart for a quick cross-reference?

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+====
+The Redpanda Helm chart's `preStop` lifecycle hook puts each broker into maintenance mode before the Pod that it's running on is terminated. If maintenance mode does not complete before the `terminationGracePeriod` the container is forcefully terminated using a `SIGKILL` command.
+
+The default `terminationGracePeriod` is 90 seconds, which should be long enough for large clusters.
+You can test different values in a development environment.
+To configure the `terminationGracePeriod`,
+use the xref:reference:redpanda-helm-spec.adoc#statefulsetterminationgraceperiodseconds[`statefulset.terminationGracePeriodSeconds`] setting.
+====
+
+. Check for topics that have a replication factor greater than one.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc:48]: I feel like we're missing a subheading before this list. I'm not quite following the flow of the guide here. We list the four top-level steps and read about going into maintenance mode. I'm not quite clear where these steps fall within the above list, if they do at all?

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+. Place a broker into maintenance mode.
+. Wait for maintenance mode to finish.
+. Terminate the Pod so that it is restarted.
+. Take the broker out of maintenance mode.
+
+When a broker is placed into maintenance mode, it reassigns its partition leadership to other brokers for all topics that have a replication factor greater than one (three is the default replication factor for topics). Reassigning partition leadership involves _draining_ leadership from the broker and _transferring_ that leadership to another broker.
+
+[NOTE]
+====
+The Redpanda Helm chart's `preStop` lifecycle hook puts each broker into maintenance mode before the Pod that it's running on is terminated. If maintenance mode does not complete before the `terminationGracePeriod` the container is forcefully terminated using a `SIGKILL` command.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]:
```suggestion
The Redpanda Helm chart's `preStop` lifecycle hook puts each broker into maintenance mode before terminating the Pod running it. If maintenance mode does not complete before the `terminationGracePeriod` the container is forcefully terminated using a `SIGKILL` command.
```
Also, should it be "...the Pod is forcefully terminated..."?

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
++
+```bash
+kubectl exec <pod-name> --namespace <namespace> -c redpanda -- \
+  rpk cluster health
+```
++
+The draining process won't start until the cluster is healthy.
+The amount of time it takes to drain a broker and reassign partition leadership depends on the number of partitions and how healthy the cluster is.
+For healthy clusters, draining leadership should take less than a minute.
+If the cluster is unhealthy, such as when a follower is not in sync with the leader, then draining the broker can take even longer.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]: Can we quantify how much longer? Maybe consider revising the wording here -- "...can take even longer." This feels like a weak statement of fact and would be stronger if we could better quantify the effects of an unhealthy cluster.

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+Healthy:                     true <1>
+Controller ID:               0
+All nodes:                   [0 1 2] <2>
+Nodes down:                  [] <3>
+Leaderless partitions:       [] <3>
+Under-replicated partitions: [] <3>
+----
+<1> The cluster is either healthy (`true`) or unhealthy (`false`).
+<2> The node IDs of all brokers in the cluster.
+<3> If the cluster is unhealthy, these fields will contain data.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]:
```suggestion
<3> These fields contain data only when the cluster is unhealthy.
```

File: modules/upgrade/pages/k-rolling-upgrade.adoc [hunk 11]
```diff
@@ -114,7 +114,7 @@ CAUTION: Make sure to include all existing overrides, otherwise the upgrade may
 [[upgrade]]
 == Perform a rolling upgrade
 
-A rolling upgrade involves running the following procedure on each broker, one at a time:
+A rolling upgrade involves running the following procedure on each broker, one at a time. The Redpanda Helm chart and the Redpanda Operator automate this procedure.
```
Comment (greg-szrama) [inline modules/upgrade/pages/k-rolling-upgrade.adoc]: Can we link to resources on the Redpanda Operator and Redpanda Helm chart for a quick cross-reference?

**Review by greg-szrama — CHANGES_REQUESTED** [review]
The flow of the guide doesn't quite make sense to me. It seems to be missing some headings between the first ordered list and the second, or some language to help indicate what the new bullets are discussing.

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc (hunk 8 above)
Comment (JakeSCahill) [reply modules/manage/pages/kubernetes/k-rolling-restart.adoc]: The grace period is the duration in seconds after the processes running in the Pod are sent a termination signal.

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc (hunk 5 above)
Comment (JakeSCahill) [reply modules/manage/pages/kubernetes/k-rolling-restart.adoc:13]: Added the rollingupdate strategy as a prereq. cc @alejandroEsc are there any more?

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc (hunk 9 above)
Comment (JakeSCahill) [reply modules/manage/pages/kubernetes/k-rolling-restart.adoc]: Thanks, I simplified this content as it doesn't provide any meaningful info for the user to act upon.

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+You must have the following:
+
+* xref:deploy:deployment-option/self-hosted/kubernetes/index.adoc[A Redpanda cluster running in Kubernetes].
+* The default xref:reference:redpanda-helm-spec.adoc#statefulset-updatestrategy-type[RollingUpdate strategy] configured in the Helm values.
+
+== What happens during a rolling restart
+
+When you run Redpanda in Kubernetes, your Redpanda cluster is managed as a StatefulSet where each broker runs inside its own Pod. As a result, you can perform a rolling restart using the Kubernetes API to terminate one Pod at a time, starting from the one with the highest ordinal.
+
+During a rolling restart the xref:deploy:deployment-option/self-hosted/kubernetes/kubernetes-production-deployment.adoc[Redpanda Helm chart] automates the following procedure on each broker, using `preStop` and `postStart` lifecycle hooks:
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]:
```suggestion
During a rolling restart the xref:deploy:deployment-option/self-hosted/kubernetes/kubernetes-production-deployment.adoc[Redpanda Helm chart] automates the following procedure on each broker, using the `preStop` and `postStart` lifecycle hooks:
```

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+. The `preStop` hook is executed immediately before a container is terminated.
+The `preStop` hook is responsible for the following:
+.. Place the broker into maintenance mode.
++
+Placing brokers into maintenance mode reduces the risk of interruption or degradation in service. When a broker is placed into maintenance mode, it reassigns its partition leadership to other brokers for all topics that have a replication factor greater than one (three is the default replication factor for topics). Reassigning partition leadership involves _draining_ leadership from the broker and _transferring_ that leadership to another broker.
+.. Terminate the Pod.
++
+After the `preStop` hook has completed its tasks, Kubernetes sends a SIGTERM signal to the container, signaling it to shut down.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]:
```suggestion
After the `preStop` hook completes its tasks, Kubernetes sends a SIGTERM signal to the container, signaling it to shut down.
```

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
++
+Maintenance mode may not have finished when the SIGTERM is sent. As a result, Kubernetes waits for the duration of the `terminationGracePeriodSeconds` for Redpanda to shut down gracefully. If it doesn't, a SIGKILL is sent to the container to forcefully terminate Redpanda.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc]:
```suggestion
Maintenance mode may not have finished when the SIGTERM is sent. As a result, Kubernetes waits for the duration of the `terminationGracePeriodSeconds` for Redpanda to shut down gracefully. If it's still executing, a SIGKILL is sent to the container to forcefully terminate Redpanda.
```

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+. The `postStart` hook is executed immediately after a container is created.
+The `postStart` hook takes the broker out of maintenance mode. This action re-integrates the broker into the cluster, allowing it to start handling requests and participate in the cluster's operations again.
+
+include::upgrade:partial$rolling-upgrades/restart-impact.adoc[leveloffset=+1]
+
+== Perform a rolling restart
+
+. Check for topics that have a replication factor greater than one.
++
+Partitions that live on only one broker will be offline during the restart. If you have topics with a replication factor of 1, and if you have sufficient disk space, temporarily xref:manage:data-migration.adoc#change-topic-replication-factor[increase the replication factor] to limit outages for these topics during the rolling upgrade.
```
Comment (greg-szrama) [inline modules/manage/pages/kubernetes/k-rolling-restart.adoc:50]:
```suggestion
Partitions living on only one broker will remain offline during the restart. If you have topics with a replication factor of 1, and if you have sufficient disk space, temporarily xref:manage:data-migration.adoc#change-topic-replication-factor[increasing the replication factor] will limit outages for these topics during the rolling upgrade.
```

File: modules/upgrade/pages/k-rolling-upgrade.adoc [hunk 16]
```diff
@@ -114,21 +115,11 @@ CAUTION: Make sure to include all existing overrides, otherwise the upgrade may
...
-A rolling upgrade involves running the following procedure on each broker, one at a time:
-
-. Upgrade the version of Redpanda on the broker that's running in the Pod with the highest ordinal.
-. Place the broker into maintenance mode.
-. Wait for maintenance mode to finish.
-. Terminate the Pod so that it is recreated with the new version of Redpanda.
-. Take the broker out of maintenance mode.
-
-Placing brokers into maintenance mode ensures a smooth upgrade of your cluster while reducing the risk of interruption or degradation in service. When a broker is placed into maintenance mode, it reassigns its partition leadership to other brokers for all topics that have a replication factor greater than one (three is the default replication factor for topics). Reassigning partition leadership involves _draining_ leadership from the broker and _transferring_ that leadership to another broker.
+A rolling upgrade involves updating the version of Redpanda on each broker and performing a rolling restart.
```
Comment (greg-szrama) [inline modules/upgrade/pages/k-rolling-upgrade.adoc]: Just to confirm I'm reading this right -- you update each Redpanda broker while it's running, then you do a rolling restart which activates the new version on each broker as it comes back online?

**Review by greg-szrama — APPROVED** [review]
A few other suggestions and one question for clarification, but this is looking good!

File: modules/upgrade/pages/k-rolling-upgrade.adoc (hunk 16 above)
Comment (JakeSCahill) [reply modules/upgrade/pages/k-rolling-upgrade.adoc]:
```suggestion
A rolling upgrade involves updating the version of Redpanda in the Redpanda  Helm chart, which triggers a rolling restart.
```

File: modules/manage/pages/kubernetes/k-rolling-restart.adoc (hunk 5 above)
Comment (alejandroEsc) [reply modules/manage/pages/kubernetes/k-rolling-restart.adoc:13]: that should be all you need. k8s, and sts with that strategy.

File: modules/upgrade/pages/k-rolling-upgrade.adoc [hunk 17]
```diff
@@ -114,21 +115,11 @@ CAUTION: Make sure to include all existing overrides, otherwise the upgrade may
...
-A rolling upgrade involves running the following procedure on each broker, one at a time:
-
-. Upgrade the version of Redpanda on the broker that's running in the Pod with the highest ordinal.
-. Place the broker into maintenance mode.
-. Wait for maintenance mode to finish.
-. Terminate the Pod so that it is recreated with the new version of Redpanda.
-. Take the broker out of maintenance mode.
-
-Placing brokers into maintenance mode ensures a smooth upgrade of your cluster while reducing the risk of interruption or degradation in service. When a broker is placed into maintenance mode, it reassigns its partition leadership to other brokers for all topics that have a replication factor greater than one (three is the default replication factor for topics). Reassigning partition leadership involves _draining_ leadership from the broker and _transferring_ that leadership to another broker.
+A rolling upgrade involves updating the version of Redpanda in the Redpanda  Helm chart, which triggers a rolling restart.
```
Comment (alejandroEsc) [inline modules/upgrade/pages/k-rolling-upgrade.adoc]:
```suggestion
A rolling upgrade involves updating the version of Redpanda in the Redpanda Helm chart, which triggers a rolling restart.
```

**Review by alejandroEsc — APPROVED** [review]
one item, otherwise lgtm. Thank you!

**Pull request merged**